relay-cogs/src/cogs.rs

```diff
@@ -0,0 +1,336 @@
+            }
+
+            let ratio = (weight.get() as f32 / total_weight as f32).clamp(0.0, 1.0);
+            Some((feature, ratio))
+        })
+    }
+}
+
+impl fmt::Debug for AppFeatures {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "AppFeatures(")?;
+
+        let mut first = true;
+        for (feature, weight) in self.weights() {
+            if !first {
+                first = false;
+                write!(f, ", ")?;
+            }
+            write!(f, "{feature:?}={weight:.2}")?;
+        }
+        write!(f, ")")
+    }
+}
+
+impl From<AppFeature> for AppFeatures {
+    fn from(value: AppFeature) -> Self {
+        Self::new(value)
+    }
+}
+
+/// A builder for [`AppFeatures`] which can be used to configure different weights per [`AppFeature`].
+pub struct AppFeaturesBuilder(AppFeatures);
+
+impl AppFeaturesBuilder {
+    /// Increases the `weight` of an [`AppFeature`].
+    pub fn add_weight(mut self, feature: AppFeature, weight: usize) -> Self {
+        let Some(weight) = NonZeroUsize::new(weight) else {
+            return self;
+        };
+
+        if let Some(previous) = self.0 .0.get_mut(feature) {
+            *previous = previous.saturating_add(weight.get());
+        } else {
+            self.0 .0.insert(feature, weight);
+        }
+
+        self
+    }
+
+    /// Sets the specified `weight` for an [`AppFeature`].
+    pub fn weight(mut self, feature: AppFeature, weight: usize) -> Self {
+        if let Some(weight) = NonZeroUsize::new(weight) {
+            self.0 .0.insert(feature, weight);
+        } else {
+            self.0 .0.remove(feature);
+        }
+        self
+    }
+
+    /// Builds and returns the [`AppFeatures`].
+    pub fn build(self) -> AppFeatures {
+        self.0
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{collections::HashMap, time::Duration};
+
+    use super::*;
+    use crate::test::TestRecorder;
+
+    #[test]
+    fn test_cogs_simple() {
+        let recorder = TestRecorder::default();
+        let cogs = Cogs::new(recorder.clone());
+
+        drop(cogs.timed(ResourceId::Relay, AppFeature::Spans));
+
+        let measurements = recorder.measurements();
+        assert_eq!(measurements.len(), 1);
+        assert_eq!(measurements[0].resource, ResourceId::Relay);
+        assert_eq!(measurements[0].feature, AppFeature::Spans);
+    }
+
+    #[test]
+    fn test_cogs_multiple_weights() {
+        let recorder = TestRecorder::default();
+        let cogs = Cogs::new(recorder.clone());
+
+        let f = AppFeatures::builder()
+            .weight(AppFeature::Spans, 1)
+            .weight(AppFeature::Transactions, 1)
+            .weight(AppFeature::MetricsSpans, 0) // Noop
+            .add_weight(AppFeature::MetricsSpans, 1)
+            .weight(AppFeature::Transactions, 0) // Reset
+            .build();
+        {
+            let _token = cogs.timed(ResourceId::Relay, f);
+            std::thread::sleep(Duration::from_millis(50));
```

Comment:
Should we use tokio's sleep instead? IIUC tokio should handle the magic of knowing when to sleep and when to update other clocks, potentially increasing test speed.

Reply:
this is not a tokio test with `async` fn